docs/testing.md

```diff
@@ -0,0 +1,212 @@
+
+### Tests Should Mirror How Software is Used
+
+One possible way to write tests is to manipulate the internal state directly and artificially increase test coverage. Unfortunately, these tests are ephemeral and brittle because they are coupled with implementation details. In other words, if a piece of code is altered or refactored, the tests need to be substantially rewritten, and we can't conclusively determine whether or not functionality has changed.
+
+The goal of tests is precisely to improve the quality of the implementation while keeping functionality constant. When testing the UI, the test might include inspecting the text in DOM elements and simulating clicking behaviors. [This PR includes good examples](https://github.com/MetaMask/metamask-extension/pull/17360/files) of that. Instead of changing the internal state and props of the component programmatically, the testing API encourages interaction with DOM elements. This was a [deliberate design choice](https://testing-library.com/docs/guiding-principles) for React Testing Library.
+
+For unit tests that are not related to the UI of the extension, we can't rely on the library to direct us on how to write our tests, but the same principle can be generalized. We aim to test a unit code for all the possible interaction scenarios. Mocks should be used sparingly, as they often make the test diverge from how the unit is used in production.
+
+### The Anatomy of a Test
+
+A good way to reason about tests is to think of their typical structure.
+
+The first block is the SETUP, where the object of the test is instantiated or rendered. This typically includes initializing the state to be passed as an argument and mocking or stubbing functions as needed.
+
+ACT is the trigger after which the functionality should occur. (The first part of "**When foo**, then bar")
+
+ASSERT is the consequence of the trigger that has happened. (The second part of "When foo, **then bar**")
+
+The TEARDOWN block typically cancels the setup by cleaning up the state to ensure test isolation (see section "Isolation and independence").
+
+You can read more about [this structure here](https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/testingandquality/aaa.md).
+
+Jest provides [APIs](https://jestjs.io/docs/setup-teardown) to remove duplication on setup and teardown.
+
+### Isolation & Independence
+
+Tests should not depend on each other to pass. This can be verified by running one test of the suite only or by changing the test order. Setup and tear-down steps are used to ensure this.
+
+You can read more on the topic [here](https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/testingandquality/avoid-global-test-fixture.md).
+
+### Manage Test Data
+
+Test data makes up for a lot of the boilerplate for tests.
+
+To avoid duplication hell and enhance readability, it's sometimes helpful to transform the base objects into object generators that accept overrides.
```

Comment:
I think here we can also add to put the data in the mock folder

Reply:
This is an initial document, but you make good points, and I created a new issue to track additional topics to be added to the documentation: https://github.com/MetaMask/metamask-extension/issues/17680